### ISSUE TITLE
[$1000] After completing "Request money" journey the chat randomly redirects to other chat Reported by @mananjadhav

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#6912]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app 
2. Have multiple chat discussions and groups 
3. Pick the top discussion on LHN 
4. Select Request money from "+" menu 
5. Select same user as your current open chat and send request 

## Expected Result:
User should stay on the same chat 

## Actual Result:
User is redirected randomly to another chat 

## Workaround:
Unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.23-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43995119/147601395-82f510e8-953d-43ab-a759-e5dd996c0e6b.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mananjadhav
**Slack conversation:**  https://expensify.slack.com/archives/C01GTK53T8Q/p1640620909132300

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Gonals (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Christinadobrzyn (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Gonals** [comment]
Setting tags and sending to the pool!

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @iwiznia (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mallenexpensify** [comment]
Job posted https://www.upwork.com/jobs/~01b794160880fbf187
Will keep 👀 on for @Christinadobrzyn til she's back (her OOO isn't showing up correctly so she got assigned to this)

**Comment by dklymenk** [comment]
Hi, I managed to narrow the issue down to this line:
https://github.com/Expensify/App/blob/615f5b3ff21cbdcf83ef2a050c875de2a0dbfb26/src/libs/Navigation/CustomActions.js#L108

I couldn't figure out the actual root cause, but for some reason dispatching a `CLOSE_DRAWER` action sometimes gets followed up with  a `RESET` action with a `reportID` for a wrong chat. I only managed to get that information with [useReduxDevToolsExtension](https://reactnavigation.org/docs/devtools/#usereduxdevtoolsextension)

![DeepinScreenshot_select-area_20220101150637](https://user-images.githubusercontent.com/64093836/147850213-8c9f4806-a87f-4813-9768-ebfb10ff5ee6.png)

I wasn't able to put that information to good use and I haven't found what exactly causes that follow-up action to get dispatched, but deleting this code that sends a `CLOSE_DRAWER` action if the user is already on the target scene fixes the issue:
https://github.com/Expensify/App/blob/615f5b3ff21cbdcf83ef2a050c875de2a0dbfb26/src/libs/Navigation/CustomActions.js#L102-L109

I understand that it's far from optimal solution, so hopefully someone manages to find a better and I hope my findings will help with that.

Thanks.

**Comment by parasharrajat** [comment]
Thanks, @dklymenk for putting an outline for the issue. I was known that this action is causing multiple issues. Currently two. I will investigate myself as I refactored it a few days back and then we cleaned up the code to make it more robust. Somehow, that cleanup has caused bugs. 

cc: @marcaaron

**Comment by K4tsuki** [comment]
#### proposal

I think the problem is `navigateBackToRootDrawer` inside `CustomActions.pushDrawerRoute`

https://github.com/Expensify/App/blob/a87b53c273be288bdc887d618973404ba01ea22b/src/libs/Navigation/CustomActions.js#L98-L100

I think we shouldn't do other navigation beside returning navigation actions in `CustomAction pushDrawerRoute`.

<br />

We could do `navigateBackToRootDrawer` right before `pushDrawerRoute` in `navigate`.
Pulling `navigateBackToRootDrawer` into `navigate`:
with additional check, something like:
```
if (isDrawerRoute(route)) {
    const windowPath = window && window.location ? window.location.pathname.substr(1) : null;
    if (isDrawerRoute(windowPath)) {
        CustomActions.navigateBackToRootDrawer();
    }
    navigationRef.current.dispatch(CustomActions.pushDrawerRoute(route));
}
```

or

```
if (isDrawerRoute(route)) {
    const rootState = navigationRef.getRootState();
    if (rootState.routes.length > 1) {
        CustomActions.navigateBackToRootDrawer();
    }
    navigationRef.current.dispatch(CustomActions.pushDrawerRoute(route));
}
```

This solve the issue for me.

**Comment by marcaaron** [comment]
>I think we shouldn't do other navigation beside returning navigation actions in CustomAction pushDrawerRoute.

Any reason for this conclusion? Definitely open to hearing proposals for how to fix this and am happy to hear we have found a solution, but we should be able to explain what is happening as well. Every time this code is modified something else breaks.

**Comment by K4tsuki** [comment]
https://reactnavigation.org/docs/navigation-prop/#advanced-api-reference

Stated: 
> The dispatch function is much less commonly used
> We recommend to avoid using the dispatch method often unless absolutely necessary.

Dispatch function:
> The dispatch method lets us send a navigation action object which determines how the navigation state will be updated

There is no clear `not allowed` or `forbidden` message on documentation on calling other navigation actions inside `dispatch`. But I think react-navigation treated it as single navigation action (or maybe only focusing on how navigation state will be updated).  
In CustomActions examples, there is no other navigation actions calling.

I am not sure, I am raising stackoverflow question about this.

**Comment by marcaaron** [comment]
Cool. Thanks for looking into it! 

To give some background here, we have done the custom navigate method to try and get various things to work for our web implementation of `react-navigation`. At this point, it has led to some pretty tricky code. So, I feel more comfortable accepting proposals that clearly explain why a fix will do what we want and show a good understanding of how the current code works. "I changed this and it works now" or "this is done incorrectly, but I can't explain why" proposals will need to be improved.

**Comment by Christinadobrzyn** [comment]
Just checking on this - @parasharrajat and @marcaaron - can you confirm we're still reviewing proposals for this job?

**Comment by parasharrajat** [comment]
Yeah @Christinadobrzyn.

**Comment by Christinadobrzyn** [comment]
Increased [Upwork job](https://www.upwork.com/jobs/~01b794160880fbf187) price to $500 - collecting proposals.

**Comment by Christinadobrzyn** [comment]
Increased job price to $1000 - collecting proposals.

**Comment by murataka** [comment]
Hello , 

The issue happens when a user request money after switching between chat members .

The bug is happening because of two reasons: 

1 :  Onyx variable lifecycles configuration seems to be incomplete for some variables.

2:  Misuse of `componentDidMount` functionality in ReportScreen.js ( should not make changes to  non-local variables at that state listener(s) )

So the solution is to complete the part related to Onyx about active (selected) report lifecycle, and then changing the lifecycle of variables in such a way that we only make changes to global variables where the trigger is the event which we can easily track: In that case it is the user interaction handler function (s) . 

Also, as I am new here , I needed to finish the debugging process to understand what is what . So I can also add a video of it working correctly in case you need it.

**Comment by parasharrajat** [comment]
Hey @murataka, I didn't understand your proposal. Could you please add more technical info about the things you are talking about? Use code references to indicate where you are suggesting the changes.

**Comment by murataka** [comment]
> Hey @murataka, I didn't understand your proposal. Could you please add more technical info about the things you are talking about? Use code references to indicate where you are suggesting the changes and what changes.

Yes, The project uses Onyx library to handle updating of multiple renderable components. Updating multiple components upon a change of a variable is a difficult task on react apps, whereas you already are using Onyx to handle that known problem.

However, when you do not add the   `withOnyx({ ... ` to a component configuration , or not add the variable to that configuration to make it read props variables from the storage , it will assume that variable is undefined. This is a case for MaindrawerNavigator.js

The current implementation tracks not just the selected report(chat session) , but also tracks when we accessed them. 
So , the implementation fetches the last used chat session (In fact it becomes previous chat session due to the lifecycle of variables ) , so the developer(s)  may got confused to using that lastaccessedreportid to store the current chat session, which leads to switching to previous instead of the last (current) chat session. This is an other case for MaindrawerNavigator.js

Also , componentdidmount and similar hook handlers should not be used for implementing pseudo algorithm to code:

we do not know when those functions will be called exactly, although we assume when they are called. For example , when a user clicks a chat button to switch between sessions , we should not implement changing a state variable , which will cause issues in some cases . It may trigger an infinite loop for example (changing session variables will cause react to render compnents which uses them ), and it will be very hard to find that kind of bug. This is the case for ReportScreen.js

As we have the Onyx library which I liked , we should be very careful when using variables from Onyx database. Those variables should be implemented carefully , and better developer should add a comment such as

```
const MainDrawerNavigator = (props) => {
// Onyx :props.reports 
    const initialParams = getInitialReportScreenParams(props.reports, !Permissions.canUseDefaultRooms(props.betas));

```

where possible. Tracking issues related to those variables will be harder after the project grows . 

The solution does not require implementing new modules , or making major changes. 

You can still ask more specific questions if you want.

**Comment by murataka** [comment]
```
const getInitialReportScreenParams = (reports, ignoreDefaultRooms) => {
    const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);

```
This is  the confusing part of the code which I mentioned above , causing the previous chat session to be fetched.

**Comment by parasharrajat** [comment]
Truly speaking, I didn't understand a single thing until https://github.com/Expensify/App/issues/6912#issuecomment-1018076712.

Maybe try to write small and clear bullet points to explain your thoughts next time. You are trying to mix suggestions with issue cause and solution. This makes it all confusing. 

 Back the topic, 
 Thanks, I see that https://github.com/Expensify/App/issues/6912#issuecomment-1018076712 is causing the wrong chat to load. 
 And I also know when it started happening. But How would you solve this issue while keeping the last accessed chat feature intact?

**Comment by murataka** [comment]
> Maybe try to write small and clear bullet points to explain your thoughts next time. You are trying to mix suggestions with issue cause and solution. This makes it all confusing.

The fix does not require much changes, it is not more than 10-20 lines of change . So , I needed to explain the cause of the problems, as otherwise I shall need to share all the changes needed. 



> Thanks, I see that [#6912 (comment)](https://github.com/Expensify/App/issues/6912#issuecomment-1018076712) is causing the wrong chat to load. I will try to test it and see if this is the main cause. And I may also know when it started happening. But How would you solve this issue while keeping the last accessed chat feature intact?

It is not the single cause and tracking that will not be enough to find the root cause. As stated above , there's a corrolation between those mistakes and all together they create the bug . 

I tried to explain the causes mostly , including the solution , whereas seeing the changes I made will make things more clear . But Instead I wanted to explain what mistakes caused that issue, and why it is difficult to debug it .

I even did not add a new variable  or function , so you can be sure that the solution is a very clear implementation ,it is not messy.

**Comment by murataka** [comment]
I will also add comments , and try to share as much details as possible , where I think that similar issues may appear and to prevent those mistakes, adding more comments to the  code seems essential to prevent future bugs.

**Comment by parasharrajat** [comment]
> I tried to explain the causes mostly , including the solution and how to prevent it happening again in future, whereas seeing the changes I made will make things more clear . But instead I wanted to explain what mistakes caused that issue, and why it is difficult to debug it .

That's great. If you have joined slack, you can start a discussion about this. And we will be glad to discuss it there in detail.

For now, let's reduce the scope of this issue to the problem it states and try to fix that first. You have explained the problem and the reason. But what I am asking is the implementation changes you would do to achieve this.

**Comment by murataka** [comment]
> For now, let's reduce the scope of this issue to the problem it states and try to fix that first. You have explained the problem and the reason. But what I am asking is the implementation changes you would do to achieve this.

I already tested the solution and it works. I would share the changes made after the contract starts, as just a few lines need to be changed :  nothing newly implemented. 

I mean it was already implemented , but there were some mistakes. As I did not introduce a new scope , there's not much to discuss, let me find and join the slack channel . I will continue to chat there . I strongly feel that the shared information above would be enough for that bug, so adding more comments here will just confuse the readers and won't be good for benefiting from the explanations/ideas shared above . See you on slack 👍

**Comment by parasharrajat** [comment]
> I already tested the solution and it works. I would share the changes made after the contract starts, 

I see. If you are worried to share your solution before the contract, then rest assured that if your solution is best of all proposed fix the issue and does not cause regressions. You would be paid for it. 

At Expensify, the process is a little different. we like to hear the solution first and analyze your approach to it. It helps us in finding the regressions and give you suggestions before you start working on the problem. It also helps us maintain visibility. 

Let me know if you still have any doubts. You can read more about the process at CONTRIBUTING.md.

**Comment by murataka** [comment]
Yes , I understand that , but just a few hours passed since I have found that repo / work conditions. I made changes to 3-4 files, some of them just 1 line maybe ., and will make more checks / try to find any mistakes if I have made before sharing all the solution ,and I planned to do that after I have a contract. 
Discussing the solution after I submit the code for review will be very nice as I am very experienced . That is because I will have more chance to defend the solution I propose or prevent any misunderstanding. 

It does not state in CONTRIBUTING.md that we must share all the solution before having the contract.

**Comment by murataka** [comment]
@parasharrajat 

Contributing.md  tells me to stop at that state ...

`5. Pause at this step until someone from the Contributor-Plus team and / or someone from Expensify provides feedback on your proposal (do not create a pull request yet).`

**Comment by parasharrajat** [comment]
Yeah, Correct. I am not asking you to create a PR rather the code changes you would do in brief.

**Comment by murataka** [comment]
@parashat , I counted 12 lines to be added ( including brackets ) and one deletion  . I have no idea if you are the one to start the contract , but I want to add this as a summary: the solution is mostly finishing the previous incomplete implementation about the chat related Onyx  library.   Still waiting for the slack invitation. Please let me know if the hiring process may take more than a week or if I need to take any other action .

**Comment by parasharrajat** [comment]
@iwiznia Could you please help me review the above proposal? Thanks.

**Comment by iwiznia** [comment]
Hi @murataka, thanks for investigating the issue!
From the comments above, it seems you identified the problem and found a solution. Our approach on all our open source issues is to first discuss the solution with the contributors in order to make sure we all agree on the solution and that it is the best solution we can find. So we do not hire anyone before we can evaluate the solution. 
I know this is in contrast to other job offers published in Upwork, but it allows us to ensure the level of quality we want for our app and code while also diminishing the work thrown out by contributors. ie: if we waited till contributors submit a PR and then found out the solution is not acceptable or has some drawbacks, then that code would need to be thrown out, which is frustrating for contributors.
So, would you mind to explain _exactly_ what changes would you do to solve this problem?

**Comment by murataka** [comment]
> So, would you mind to explain _exactly_ what changes would you do to solve this problem?

I already tried to explain the root causes of the bug , so I think I shall share the code changes need to be made , and additional to my comments above , I hope that will be more clear for you. 

### ReportScreen.js

```
  componentDidMount() {
        this.prepareTransition();
        this.storeCurrentlyViewedReport();  // This line makes changes on Onyx variables , and should not be implemented here 
 // details explained above : search the "componentDidMount" keyword
    } 
```

### SidebarLinks.js
```


onSelectRow={(option) => {
                        Report.updateCurrentlyViewedReportID(option.reportID);
// instead the deleted line should be added here , yes they are different  functions but does the same thing.
```

### MainDrawerNavigator.js

```
 reports: {
        key: ONYXKEYS.COLLECTION.REPORT,
    },
     currentlyViewedReportID: {
      key: ONYXKEYS.CURRENTLY_VIEWED_REPORTID,
        },
    betas: {

// we need to integrate ONYXKEYS.CURRENTLY_VIEWED_REPORTID at his file, that was missing, 

// the developer who implemented it just got confused because there is some similar functionality which returns previous document instead of the last one .

 

```

  ```
if (!initialParams.reportID) {
        return <FullScreenLoadingIndicator />;
    }
    if (props.currentlyViewedReportID) { /// added that if clause not to break the existing functionality. and also the answer to 
 // the question which @parasharrajat  asked above .
> But How would you solve this issue while keeping the last accessed chat feature intact?


        initialParams.reportID=props.currentlyViewedReportID;
        initialParams.currentlyViewedReportID=props.currentlyViewedReportID;
    }
```

This is pretty the whole solution,

**Comment by murataka** [comment]
in case you want to test, 

You shall also need the below lines  at MainDrawerNavigator.js :
 ```
 reports: PropTypes.objectOf(PropTypes.shape({
        reportID: PropTypes.number,
    })),
    currentlyViewedReportID: PropTypes.string,
    /** Beta features list */
    betas: PropTypes.arrayOf(PropTypes.string),
const defaultProps = {
    reports: {},
    betas: [],
    currentlyViewedReportID: '',
};

```

**Comment by murataka** [comment]
let me know if you are trying to test it , there might be missing changes still . I don't know if  this is the right way to evaluate the proposal in case you are trying to test it  .

**Comment by murataka** [comment]
ReportScreen.js

 `       const reportID = this.props.currentlyViewedReportID;
`

Please let me know if sharing the code will be mandatory for all bugfix issues for my next proposals, or if there's some way to not make that progress so long so I can take action about it ..

**Comment by murataka** [comment]
https://user-images.githubusercontent.com/5358438/150614123-7e6b89f9-4615-4cf4-9170-f9f24e73eaf4.mov

**Comment by brianmuks** [comment]
**Proposal**
``` Navigation.dismissModal();``` will fix the bug. 
We can call the method on this line: 
https://github.com/Expensify/App/blob/0bce09177d2aed8c5e377bc758f5241dd22886eb/src/pages/iou/IOUModal.js#L308
or when the transaction has succeeded.

**Comment by murataka** [comment]
> 

that time the green (+) button to request money will cause chat to switch to the report chat for that last request made. Have no idea about the intended behavior though.

**Comment by parasharrajat** [comment]
I am having a hard time putting all the pieces together. @murataka. So I would request you to write down the complete proposal. 
- Explain to me why and how you think that `this.storeCurrentlyViewedReport()` is causing the issue?
- Your proposed changes against the issue.
- Explain your changes, not just the code. 

Most of the time testing the proposal is not required.

**Comment by murataka** [comment]
> I am having a hard time putting all the pieces together. @murataka. So I would request you to write down the complete proposal.

I apologize for being complicated, that is because I am so new here, still trying to get used to. Let me try to be more clear ( As much as I can )

**Comment by brianmuks** [comment]
@parasharrajat what do you think of my solution ?

**Comment by parasharrajat** [comment]
@brianmuks Please read [this post](https://github.com/Expensify/App/issues/6912#issuecomment-1005168037) and let me know what do you think?

**Comment by murataka** [comment]
> * Explain to me why and how you think that `this.storeCurrentlyViewedReport()` is causing the issue?

This in fact might need further discussion maybe in the slack channel. Because updating the Onyx variables at the render flow is not a good idea. Example case : 

```
class ChildComponent extends Component {
...
  componentDidMount() {
Onyx.merge(ONYXKEYS.someOnyxStateVariable, String(newValueOfsomeOnyxStateVariable));
}
<ParentComponent sompeproperty={someOnyxStateVariable}>

<ChildComponent > </ChildComponent  > 
</ParentComponent >
 
```
This will recursively cause the React renderer to render ParentComponent . I did not test that , but may do a quick test if I'm right if you ask so . Also look at hist issue : https://github.com/Expensify/App/issues/5930

**Comment by brianmuks** [comment]
> @brianmuks Please read [this post](https://github.com/Expensify/App/issues/6912#issuecomment-1005168037) and let me know what do you think?

Thanks very helpful 🙏🏾 . 

So I discovered that whenever the modal is opened going back to the chats before closing the modal affects the state of the chat screen .  On the other hand closing the modal then navigating back to the chat screen maintains the state of the chat screen. I am not yet sure of the underlying cause.

**Comment by murataka** [comment]
> * Your proposed changes against the issue.
> 
> * Explain your changes, not just the code.

we are currently using this variable instead of currentlyViewedReportID
`    const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);
`
But this variable is not the same , so when we use the green (+) button to request transaction , the active chat window switches to last made transaction. Even if that is the intended behavior , we should still use currentlyViewedReportID in implementation to have more understandable code.

So , my proposal is to update SidebarLinks as

```
  onSelectRow={(option) => {
          +              Report.updateCurrentlyViewedReportID(option.reportID);
```

and  use currentlyViewedReportID instead of 

`    const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);
`
where essential . it is currently used in ReportScreen.js and SidebarLinks.js  . 

There might be some more little similar changes need to be made , I need to check further. But this is the general idea.

**Comment by parasharrajat** [comment]
@brianmuks 

> So I discovered that whenever the modal is opened going back to the chats before closing the modal affects the state of the chat screen. On the other hand, closing the modal then navigating back to the chat screen maintains the state of the chat screen. I am not yet sure of the underlying cause.
 
 First, we need this behavior to fix another bug where drafts do not update on report change. So we reset the navigation state. 
 Second, in the navigation action, first, the side modal is closed and then the drawer is loaded with the correct report. You can try to debug the navigation action and observe the same. 
 
 That post points out that ` I am not yet sure of the underlying cause.` You should be able to explain. That is a complex piece of logic so we won't do any change until fully reasoned.
 
 --- 
 
 Thanks, @murataka, this could be the issue but do you have any explanation why it works sometimes.

**Comment by murataka** [comment]
> Thanks, @murataka, this could be the issue but do you have any explanation why it works sometimes.

The bug will happen if you previously switched between conversations before making transaction. 

It switches to  that variable :

`const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);`

So it is not random.

**Comment by parasharrajat** [comment]
Ah, I see. But I think `currentlyViewedReportID` is not synced across devices so it won't help us preserve the lastViewreport functionality. 
1. User should land to the last viewed report on any device. 
2. Then he should be able to switch to any report he wants which will become last viewed. 
2. last Viewed report should only act when a user opens the app for the first time. 
3. We pass `const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);` to the initialParams and remount the drawer when screen is resized. I am not sure but I think it also remounts when we navigate to chat. 

So we need a way to achieve point 3.

**Comment by murataka** [comment]
> Ah, I see. But I think `currentlyViewedReportID` is not synced across devices so it won't help us preserve the lastViewreport functionality.

I don't think this will be an issue because I'm not replacing all code to use `const last = ReportUtils.findLastAccessedReport(reports, ignoreDefaultRooms);`

It is still being used where it should be used, so `last Viewed report should only act when a user opens the app for the first time.` won't be affected . this is why I told that may need to check further , and that comment is helpful about that.

**Comment by murataka** [comment]
> Ah, I see. But I think `currentlyViewedReportID` is not synced across devices so it won't help us preserve the lastViewreport functionality.

```
if (props.currentlyViewedReportID) {
     //   initialParams.reportID = props.currentlyViewedReportID; <- This line is wrong , won't be implemented.
        initialParams.currentlyViewedReportID = props.currentlyViewedReportID;
    }
```

**Comment by murataka** [comment]
```
const MainDrawerNavigator = (props) => {
    const initialParams = getInitialReportScreenParams(props.reports, !Permissions.canUseDefaultRooms(props.betas));

    // Wait until reports are fetched and there is a reportID in initialParams
    if (!initialParams.reportID) {
        return <FullScreenLoadingIndicator />;
    }
    if (props.currentlyViewedReportID) {
    //    initialParams.reportID = props.currentlyViewedReportID;
        initialParams.currentlyViewedReportID = props.currentlyViewedReportID;
    }

```

**Comment by murataka** [comment]
ReportScreen.js
```
  const reportID;
if(this.props.currentlyViewedReportID)
 reportID= this.props.currentlyViewedReportID;
else
reportID= this.props.reportID;
```

**Comment by murataka** [comment]
@parasharrajat  Should I prepare those changes at a commit ? This is taking some time to compare maybe ?

**Comment by parasharrajat** [comment]
I understood your proposal but I still have my own doubts. I will look into it later. Currently AFK.

**Comment by murataka** [comment]
> 

ok , I shall prepare a commit on my own clone and it will be easier to examine , also will check about componentDidMount case if it is an issue .

**Comment by murataka** [comment]
@parasharrajat  

You can check the changes here : 
https://github.com/murataka/App/tree/playground

**Comment by murataka** [comment]
For the componentdidmount issue ,
Eslint already generates that warning below, however as it is being called in a function , eslint is unable to detect that problem , so we are not getting that information currently. 

<img width="1440" alt="Screen Shot 2022-01-24 at 06 03 37" src="https://user-images.githubusercontent.com/5358438/150715326-a24982d5-1b08-4cde-a2d5-5d3ec291518e.png">

**Comment by murataka** [comment]
Will check further for componentdidmount about updating onyx data.

**Comment by parasharrajat** [comment]
I will review this thoroughly today. But I don't understand https://github.com/Expensify/App/issues/6912#issuecomment-1019627837. 
https://github.com/Expensify/App/issues/6912#issuecomment-1019631496

How will you load the lastViewed report on the Drawer? In that code you are picking the `currentlyViewedReportID`.

**Comment by murataka** [comment]
> I will review this thoroughly today. But I don't understand [#6912 (comment)](https://github.com/Expensify/App/issues/6912#issuecomment-1019627837). [#6912 (comment)](https://github.com/Expensify/App/issues/6912#issuecomment-1019631496)
> 
> How will you load the lastViewed report on the Drawer? In that code you are picking the `currentlyViewedReportID`.


I have continued to work on it , will let you know after updating it again . I think we are on the same page that we need to have more control to manage when to switch the viewedreport (active chat or whatever you say).  

I had found some more changes needed to be made , like Navigaton.navigate was called at requestmoney completion, but not sendmony (IOU implementation) . 

Bringing all together , currentlyViewedReportID was already being used and working fine for SidebarLinks.js (after setting the updatecurrentlyviewdreport to event instead of componentdidmount) . So in the end we might have something like ->

```
AppInit -> unset( currentlyViewedReportID )

 Maindrawer -> 
    ifset (currentlyViewedReportID) show currentlyViewedReportID
       else show lastReportId
    
```

**Comment by murataka** [comment]
<img width="1438" alt="Screen Shot 2022-01-25 at 05 17 58" src="https://user-images.githubusercontent.com/5358438/150898536-9bcb0d50-ba3a-4960-ac66-8ae236e9e11e.png">

No need to navigate completed requests . They already order to top , and seems userfriendly, this was already implemented that way on sendmoneyIOU

**Comment by parasharrajat** [comment]
> No need to navigate completed requests . They already order to top at left sidebar, and seems userfriendly, this was already implemented that way on sendmoneyIOU

How to did you end up with the conclusion that they are already on the top. When we request money from some user from the Fab button, how will you navigate the user to that Chat.

**Comment by murataka** [comment]
> How to did you end up with the conclusion that they are already on the top.

This functionalities can now be managed as I finished my changes I propose, so we can discuss what the behavior should be . As this issue is related to "request money" only , I tried to make it work similar to the other functionalities like sendMoney.

You should be able to test and compare from here : 

https://github.com/murataka/App/commits/fix-wrongreportfetch

**Comment by murataka** [comment]
ps: I also hope this fix may solve some other issues, so if you conclude that the solution is not bad, we should also test any related issue like https://github.com/react-navigation/react-navigation/pull/10262

**Comment by murataka** [comment]
Onyx.merge(ONYXKEYS.CURRENTLY_VIEWED_REPORTID, String(''));

should be called after app initialization , so the below commit won't clear it each time , but I did not yet found where it should be best to clear it yet. https://github.com/murataka/App/commit/ee9565ae880097d18fddf22a18c1abcba2109abe

**Comment by murataka** [comment]
also , https://github.com/Expensify/App/issues/7350 seems to be solved after applying above proposed changes.

**Comment by parasharrajat** [comment]
Sorry for delay. I will wrap this up today. I have one doubt only.

**Comment by murataka** [comment]
> Sorry for delay. I will wrap this up today. I have one doubt only.

Also let me know if evaluating actual proposed code + the explanations will be nice for evaluating proposals. I really don't want to flood the slack and github with my messages, and would like to be more efficiently communicating.

**Comment by parasharrajat** [comment]
> Also let me know if evaluating actual proposed code + the explanations will be nice for evaluating proposals. I really don't want to flood the slack and github with my messages, and would like to be more efficiently communicating.

+1 for this. I would suggest that you combine your message into one post.

**Comment by murataka** [comment]
for this one , you can evaluate 

https://github.com/murataka/App/commits/fix-wrongreportfetch

not mandatory to read whole posts again and again.

**Comment by parasharrajat** [comment]
Thanks, I was looking into the same. I tried to debug the cause yesterday. I may have found some other issues but I am still evaluating your solution. So far, I think it will cause issues but let's see.

**Comment by murataka** [comment]
> Thanks, I was looking into the same. I tried to debug the cause yesterday. I may have found some other issues but I am still evaluating your solution. So far, I think it will cause issues but let's see.

removing Navigator.navigate methods from requestmony IOU ( https://github.com/murataka/App/commit/f5e5a72bd6a499339e96ed0ecda488fbccde0a6e )

 
would solve the issue when we want to focus the bugs only and only about this topic  . I can move the rest to some other navigation related issues, because I solved some bugs unintentionally while trying this one. That might be confusing while evaluating this proposal ( like: "hey why did you need to change that ") ...

**Comment by murataka** [comment]
Proposal

removing Navigator.navigate methods from requestmoney IOU ( murataka@f5e5a72 )

fixes the related bug , and it would be easier to evaluate this proposal. 

So I created this proposal again which is just focusing on this bug to make things simler.

What it does is just remove the 

`navigation.navigate`

methods which are causing the actual bug.

The rest is linter removing unused code.

As the title says
> After completing "Request money" journey the chat randomly redirects to other chat 

This is the solution.

**Comment by parasharrajat** [comment]
No, Update. Still holding...

**Comment by parasharrajat** [comment]
@Expensify/applause Is this issue still reproducible? If yes, can I get fresh reproduction steps?

**Comment by JalalMitali** [comment]
I just tried and I cant see this problem I have 3 chats :)

**Comment by parasharrajat** [comment]
I have researched this one over the weekend and I have some theory but I am collecting some data to validate it. Overall, I was not able to reproduce this issue since the start but I have a feeling that it exists. I have asked for new steps and QA to retest it.

**Comment by isagoico** [comment]
We were able to reproduce this on our side on Mac 10.15.7 / Chrome Version 98.0.4758.80
1. Request money from user 1
2. Repeat request with different user 2
After second request user redirect to user 1 conversation. 

https://user-images.githubusercontent.com/44479856/154994723-bcaa79f2-0c56-4237-9daa-42be03262f4e.mp4

**Comment by parasharrajat** [comment]
Waiting for some data from the team before running some tests...

**Comment by michaelhaxhiu** [comment]
So this issue is still un-assigned right now as far as I can see, right? Mind providing a link to what we are waiting on right now @parasharrajat?

I'm trying to discern if this hold should be on hold for some reason, or doubling in value this week.

**Comment by parasharrajat** [comment]
I don't feel this needs doubling this week. I will update you next week on the status. I haven't completed my testing and it's not on hold now. I am debugging this. I have some data but am not yet ready to be shared.

**Comment by michaelhaxhiu** [comment]
Thanks for sharing the context, that's helpful!

**Comment by parasharrajat** [comment]
Ok, there is really something going on but unfortunately, I am not able to reproduce it. So back to proposals. 

I don't like @murataka 's proposal. You are suggesting to move away from  `navigation.navigate` to navigate to report in this case. And rely on `currentlyViewedReportID` to manage it. I don't think this is a good approach.

**Comment by murataka** [comment]
@parasharrajat , thanks for examining the proposals , In fact, This issue was my first and I had trouble to getting used to here at first, so , me trying to handle more than one case above is misleading. 

The only change I suggest is this one : 

https://github.com/murataka/App/commit/f5e5a72

currentlyViewedReportID is not related to that one . 

What https://github.com/murataka/App/commit/f5e5a72  does is , we are removing the unessential navigate methods. After that , it works because , we are not redirecting to wrong report, even we are not redirecting all in all .

**Comment by parasharrajat** [comment]
Thanks for clarifying the changes. There are two ways Request Money works: 
1. Clicking :heavy_plus_sign: FAB button. 
2. From the active report, :heavy_plus_sign: on the Composer. 

I think your proposal may work while using 2 but it won't when using 1. Am I missing something?

**Comment by murataka** [comment]
Let me check it and confirm. I shall pull main again and test, also add the video recording ...

**Comment by murataka** [comment]
To clarify what changed : 

We just delete those two lines. After that , we shall have unused variables and libraries. 

https://github.com/murataka/App/commit/f5e5a72#r67467635

https://github.com/murataka/App/commit/f5e5a72#r67467659

What we get in last total pr is the above lines deleted + unused variables and libraries removed.

https://user-images.githubusercontent.com/5358438/155732301-eecd86b1-043d-486f-ab6f-284023f9c7e0.mov

**Comment by parasharrajat** [comment]
This is open for proposals...

**Comment by mallenexpensify** [comment]
I'm unable to reproduce and Rajat mentioned about he's not able to either on Desktop nor staging/web/Chrome. 
@mananjadhav and/or @mvtglobally , can you see if you can reproduce? If not, we can close this

**Comment by mallenexpensify** [comment]
Gonna close this.  If anyone can reproduce, please add a comment with steps and a video screenshot, then we'll reopen
